Source/Diagnostics/ComputeDiagFunctors/BackTransformParticleFunctor.cpp

```diff
@@ -0,0 +1,175 @@
+
+                auto index = std::make_pair(pti.index(), pti.LocalTileIndex());
+
+                const auto GetParticleFilter = SelectParticles(pti, tmp_particle_data,
+                                               m_current_z_boost[i_buffer],
+                                               m_old_z_boost[i_buffer]);
+                const auto GetParticleLorentzTransform = LorentzTransformParticles(
+                                                         pti, tmp_particle_data,
+                                                         t_boost, dt,
+                                                         m_t_lab[i_buffer]);
+
+                long const np = pti.numParticles();
+
+                FlagForPartCopy.resize(np);
+                IndexForPartCopy.resize(np);
+
+                int* const AMREX_RESTRICT Flag = FlagForPartCopy.dataPtr();
+                int* const AMREX_RESTRICT IndexLocation = IndexForPartCopy.dataPtr();
+
+                const auto& ptile_src = particles.at(index);
+                auto src_data = ptile_src.getConstParticleTileData();
+                // Flag particles that need to be copied if they cross the z-slice
+                // setting this to 1 for testing (temporarily)
+                amrex::ParallelFor(np,
+                [=] AMREX_GPU_DEVICE(int i)
+                {
+                    Flag[i] = GetParticleFilter(src_data, i);
+                });
+
+                const int total_partdiag_size = amrex::Scan::ExclusiveSum(np,Flag,IndexLocation);
+                auto& ptile_dst = pc_dst.DefineAndReturnParticleTile(lev, pti.index(), pti.LocalTileIndex() );
+                auto old_size = ptile_dst.numParticles();
+                ptile_dst.resize(old_size + total_partdiag_size);
+                auto count = amrex::filterParticles(ptile_dst, ptile_src, GetParticleFilter, 0, old_size, np);
+                auto dst_data = ptile_dst.getParticleTileData();
+                amrex::ParallelFor(np,
+                [=] AMREX_GPU_DEVICE(int i)
+                {
+                   if (Flag[i] == 1) GetParticleLorentzTransform(dst_data, src_data, i,
+                                                                 old_size + IndexLocation[i]);
+                });
+                total_particles_added += count;
+            }
+        }
+        amrex::Gpu::synchronize();
+    }
+    totalParticleCounter = pc_dst.TotalNumberOfParticles();
+}
+
+
+void
+BackTransformParticleFunctor::InitData()
+{
+    m_current_z_boost.resize(m_num_buffers);
+    m_old_z_boost.resize(m_num_buffers);
+    m_t_lab.resize(m_num_buffers);
+    m_perform_backtransform.resize(m_num_buffers);
+}
+
+void
+BackTransformParticleFunctor::PrepareFunctorData ( int i_buffer, bool z_slice_in_domain,
+                              amrex::Real old_z_boost, amrex::Real current_z_boost,
+                              amrex::Real t_lab, int snapshot_full)
+{
+    m_old_z_boost[i_buffer] = old_z_boost;
```

Comment:
Let's replace those with `.at(...)` calls, so we get exceptions when we accidentally access out-of-bounds.